PWGDQ/Core/MixingHandler.cxx

```diff
@@ -0,0 +1,194 @@
+  fNMixingVariables++;
+  VarManager::SetUseVariable(var);
+}
+
+//_________________________________________________________________________
+void MixingHandler::AddMixingVariable(int var, int nBins, std::vector<float> binLims, TString varName)
+{
+
+  float* bins = new float[nBins];
+  for (int i = 0; i < nBins; ++i) {
+    bins[i] = binLims[i];
+  }
+  AddMixingVariable(var, nBins, bins, varName);
+}
+
+//_________________________________________________________________________
+int MixingHandler::GetMixingVariable(TString vars)
+{
+  int varNum = -1;
+  for (int iVar = 0; iVar < fNMixingVariables; ++iVar) {
+    if (!fVariableNames[iVar].CompareTo(vars)) {
+      varNum = iVar;
+    }
+  }
+  return varNum;
+}
+
+//_________________________________________________________________________
+std::vector<float> MixingHandler::GetMixingVariableLimits(TString vars)
+{
+  std::vector<float> binLimits;
+  for (int iVar = 0; iVar < fNMixingVariables; ++iVar) {
+    if (!fVariableNames[iVar].CompareTo(vars)) {
+      for (int iBin = 0; iBin < fVariableLimits[iVar].GetSize(); ++iBin) {
+        binLimits.push_back(fVariableLimits[iVar].At(iBin));
+      }
+    }
+  }
+  return binLimits;
+}
+
+//_________________________________________________________________________
+void MixingHandler::Init()
+{
+  //
+  // Initialization of pools
+  // NOTE: The master array holding tracks is a 1D array but it will be represented as an n-dim array
+  //       The size of the array will be N_1 x N_2 x ... x N_n
+  //       The correct event category will be retrieved using the function FindEventCategory()
+  //
+  int size = 1;
+  for (int iVar = 0; iVar < fNMixingVariables; ++iVar) {
+    size *= (fVariableLimits[iVar].GetSize() - 1);
+  }
+
+  fIsInitialized = kTRUE;
+}
+
+//_________________________________________________________________________
+int MixingHandler::FindEventCategory(float* values)
+{
+  //
+  // Find the event category corresponding to the centrality, vtxz and ep values
+  //
+  if (fNMixingVariables == 0) {
+    return -1;
+  }
+  if (!fIsInitialized) {
+    Init();
+  }
+
+  int bin[kNMaxVariables];
+  for (int i = 0; i < fNMixingVariables; ++i) {
+    bin[i] = TMath::BinarySearch(fVariableLimits[i].GetSize(), fVariableLimits[i].GetArray(), values[fVariables[i]]);
+    if (bin[i] == -1 || bin[i] == fVariableLimits[i].GetSize() - 1) {
+      return -1; // all variables must be inside limits
+    }
+  }
+
+  int category = 0;
+  for (int iVar = 0; iVar < fNMixingVariables; ++iVar) {
+    int tempCategory = 1;
+    for (int iVar2 = iVar; iVar2 < fNMixingVariables; ++iVar2) {
+      if (iVar2 == iVar) {
+        tempCategory *= bin[iVar2];
+      } else {
+        tempCategory *= (fVariableLimits[iVar2].GetSize() - 1);
+      }
+    }
+    category += tempCategory;
+  }
+  return category;
+}
+
+//_________________________________________________________________________
+int MixingHandler::GetBinFromCategory(int iVar, int category) const
```

Comment:
This is a public interface function, and the user might not know the internal representation of the variables. Here iVar would have to be the exact index of a given variable as used in the calculation of the event category. I think it would be more useful to have a function which would have the form: GetBinFromCategory(VarManager::Variables var, int category). Also, for the user it may be more useful to get information on the range of the variable corresponding to this category.
In the AliPhysics, this function was mainly used as an internal utility for the mixing handler, with not much use externally. But with the interface i suggested, it could be used in a loop over mixing categories and retrieve info on variable ranges, etc.

Reply:
This is why I added "names" for the variables. Because I wanted to access the internal representation of the variables, which I do not know prior. Using a VarManger::variable is a much more elegant way to do it!